keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+   design of GC after use (part 2)][].
+
+### Risks and Mitigations
+
+The changes to garbage collector will affect the whole system and need to be
+carefully tested. Tests including, but not limited to:
+
+* Basic tests for each of the garbage collection features:
+  * Resources can be garbage collected.
+  * Resources that opt-out of garbage collection shouldn’t be deleted.
+* When this feature is enabled and disabled, resources won’t be garbage
+  collected.
+* All garbage collection mechanisms can work coherently.
+* Works in clusters with local timezone that changes +/-1hr for DST. 
+
+This feature will also be feature gated.
+
+## Graduation Criteria
+
+This can be promoted to beta when it satisfies users' need for cleaning up
+resources after use, without regressions.
+
+This will be promoted to GA once it's gone a sufficient amount of time as beta
+with no changes. 
+
+## Implementation History
+
+TBD
+
+## Future Work
+
+### Generic Capacity-based Garbage Collection
+
+We may potentially want to support generic capacity-based garbage collection.
+Generic capacity-based garbage collection will allow users to set a cluster-wide
+threshold on each of Kubernetes resources, and garbage collect resources after
+the threshold is hit. Note that only resources that are no longer in use can be
+cleaned up. 
+
+Points to consider when implementing the capacity-based garbage collection:
+
+* Fairness: for example, the capacity can be shared equally among namespaces.
+* Resource quota: capacity should be supplementary to resource quotas and can be
+  set according to resource quota.
+* Be generic and configurable.
+* Remove resources based on their last used time, instead of creation time.
+* Be coherent with other existing garbage collection mechanism.
+
+### Namespace-wide Garbage Collection Policy
+
+We can implement a garbage collection policy that applies to all resources in
+the same namespace by default for easier garbage collection policy management.
+
+### Move Event TTL out of etcd TTL
+
+We may consider moving Event TTL out of etcd TTL, and use the TTL mechanism in
+garbage collector for Events. 
+
+### Make Pod GC and Event GC Work Better with Garbage Collector
+
+Pod GC and Event GC don't respect OwnerReference-based garbage collector today.
+The biggest problem caused by Pod GC is that it breaks Job controller
+([#28486][]).
+
+We may move Pod GC and Event GC into garbage collector, or deprecate them and
+delete them entirely.
+
+### Move deletionGracePeriodSeconds from metadata to Pod spec
+
+`deletionGracePeriodSeconds` is currently in `.metadata` of all resources,
+although it only applies to Pods. This is confusing to the users.
+
+Because users cannot set or modify this field on a non-Pod resource (the field
+will be silently dropped), it will be compatible to move it outside of
+`.metadata` of resources other than Pod. This field can be part of a
+Pod-specific `.metadata`, or can even move this field to Pod spec in the future
```

Comment:
`.metadata` attributes across various types of resources should be as similar as possible. Hence, I think it makes more sense to move `deletionGracePeriodSeconds` field to pod spec in future.

Reply:
Yes, but it has to be v2 Pod for compatibility.